packages/venia-concept/src/ServiceWorker/Utilities/htmlHandler.js

```diff
@@ -1,7 +1,144 @@
 import { HTML_UPDATE_AVAILABLE } from '@magento/venia-ui/lib/constants/swMessageTypes';
+import { parse } from 'node-html-parser';
 
 import { sendMessageToWindow } from './messageHandler';
 
+const generateScriptResource = htmlElement => htmlElement.attributes.src;
+
+const generateLinkResource = htmlElement => htmlElement.attributes.href;
+
+const generateStyleResource = htmlElement => htmlElement.text;
+
+const identity = x => x;
+
+/**
+ * Generates list of resources from the parsed HTML object.
+ *
+ * @param {HTMLElement} parsedHTML
+ *
+ * @returns { scripts: [string], links: [string], styles: [string] }
+ */
+const generateResources = parsedHTML => {
+    const resources = {};
+
+    resources.scripts = parsedHTML
+        .querySelectorAll('script')
+        .map(generateScriptResource)
+        .filter(identity)
+        .sort();
+
+    resources.links = parsedHTML
+        .querySelectorAll('link')
+        .map(generateLinkResource)
+        .filter(identity)
+        .sort();
+
+    resources.styles = parsedHTML
+        .querySelectorAll('style')
+        .map(generateStyleResource)
+        .filter(identity)
+        .sort();
+
+    return resources;
+};
+
+/**
+ * Returns true if resouce url/urls have changed.
+ *
+ * @param {HTMLElement} newDOM
+ * @param {HTMLElement} oldDOM
+ *
+ * @returns Boolean
+ */
+const haveResourcesChanged = (newDOM, oldDOM) => {
+    const newResources = generateResources(newDOM);
+    const oldResources = generateResources(oldDOM);
+    const keysToCompare = Object.keys(newResources);
+
+    return !keysToCompare.reduce((acc, key) => {
+        const hasResourceChanged =
+            newResources[key].toString() === oldResources[key].toString();
```

Comment:
What happens if one of the old resources is dropped entirely? This should be rewritten to check for removal of old keys, addition of new keys, and equivalence of unchanged values.

Reply:
If an old resource is removed the condition will fail and `hasResourceChanged` will be set to true and the page will be asked to refresh. Isn't that how it is supposed to be?

Reply:
You're going to run into errors at runtime right?

Reply:
> ~~If an old resource is removed, it won't cause a refresh.~~ If a new resource is added, then it will cause the page to refresh by setting `hasResourceChanged`. ~~Thought i realized there will be a `TypeError` in such cases. Will be fixing it in the next commit.~~

I would like to take a couple things back.

1. If a resource/resources are removed or added, page will refresh. This is already being done. `newResources` and `oldResources` are not a list, they are of type `{scripts: [], links: [], styles: []}`. If a resource/resources are added or removed, their respective array changes and a refresh will be triggered. 

2. There won't be any `TypeErrors`. The code works as-is. Though I will be adding a couple more tests to replicate this behavior.

Reply:
Added tests to replicate the behavior.